lib/ddtrace/contrib/sinatra/header_tagger_middleware.rb

```diff
@@ -0,0 +1,72 @@
+require 'ddtrace/contrib/sinatra/request_span'
+
+module Datadog
+  module Contrib
+    module Sinatra
+      # Middleware used for automatically tagging configured headers and handle request span
+      class HeaderTaggerMiddleware
+        def initialize(app)
+          @app = app
+        end
+
+        def call(env)
+          span = RequestSpan.span!(env)
+
+          # Request headers
+          parse_request_headers(env).each do |name, value|
+            span.set_tag(name, value) if span.get_tag(name).nil?
+          end
+
+          status, headers, response_body = @app.call(env)
+
+          # Response headers
+          parse_response_headers(headers).each do |name, value|
```

Comment:
Can you confirm if this is the call order?

```
Middleware.call(env)
  RequestSpan.span!(env)
    Datadog::Tracer.tracer.trace('sinatra.request')
  parse_request_headers(env)
    span.set_tag
  @app.call(env)
    app.before
    app.after
      span.finish
  parse_response_headers(env)
    span.set_tag
```

If so, I think there's a problem with having a `span.set_tag` call after we do `span.finish`. We shouldn't modify spans that have already been finished, because they might be immediately flushed by the writer. All modifications to the span should happen before its closed.

My proposal would be to change this `HeaderTaggerMiddleware` into `TracerMiddleware` that is responsible for both starting and finishing the span. Then we should refactor as much of the behavior as we can from the `Tracer` into this `Middleware` instead. For whatever can't move into this middleware (e.g. setting the resource as the route, tags, etc), the existing `Tracer` extension can instead access & modify the span via `env`, like you've had it doing already.

Beyond solving the existing span start/finish problem, what I like about this is that the `Tracer` extension effectively becomes an extension to this middleware, and the middleware is no longer co-dependent on the `Tracer` extension to be in place. It also means Sinatra app users could, in theory, extend or otherwise modify the `Tracer` module to introduce some custom modifications to the Sinatra tracing.

What do you think?

Reply:
Yeah it looks like `TracerMiddleware` is better name when it handles whole lifecycle of the span. 

As for old `Tracer` its still there to fill in Sinatra provided data that is not otherwise available from Rack context. Currently both do not depend on each other. I.e. they will work independently but to get good span information both should run.

I don't think there is an alternative approach other than not using `Rack Middleware` and not being able to tag every header in the response.